File: src/weaponsprocurement/gui/WimGuiButtonBinding.java
package weaponsprocurement.gui;

import com.fs.starfarer.api.ui.ButtonAPI;
import com.fs.starfarer.api.ui.CustomPanelAPI;
import com.fs.starfarer.api.ui.PositionAPI;

final class WimGuiButtonBinding<A> {
    private final CustomPanelAPI panel;
    private final ButtonAPI button;
    private final A action;

    WimGuiButtonBinding(CustomPanelAPI panel, ButtonAPI button, A action) {
        this.panel = panel;
        this.button = button;
        this.action = action;
    }

    boolean consumeIfPressed() {
        if (button == null || !button.isChecked()) {
            return false;
        }
        clear();
        return true;
    }

    boolean consumeIfClicked(float x, float y) {
        if (panel == null || panel.getPosition() == null || !contains(panel.getPosition(), x, y)) {
            return false;
        }
        clear();
        return true;
    }

    void clear() {
        if (button != null) {
            button.setChecked(false);
        }
    }

    A getAction() {
        return action;
    }

    private static boolean contains(PositionAPI position, float x, float y) {
        return x >= position.getX()
                && x <= position.getX() + position.getWidth()
                && y >= position.getY()
                && y <= position.getY() + position.getHeight();
    }
}

File: src/main/kotlin/weaponsprocurement/gui/WimGuiButtonBinding.kt
package weaponsprocurement.gui

import com.fs.starfarer.api.ui.ButtonAPI
import com.fs.starfarer.api.ui.CustomPanelAPI
import com.fs.starfarer.api.ui.PositionAPI

class WimGuiButtonBinding<A>(
    private val panel: CustomPanelAPI?,
    private val button: ButtonAPI?,
    private val action: A,
) {
    fun consumeIfPressed(): Boolean {
        if (button == null || !button.isChecked) {
            return false
        }
        clear()
        return true
    }

    fun consumeIfClicked(x: Float, y: Float): Boolean {
        if (panel == null || panel.position == null || !contains(panel.position, x, y)) {
            return false
        }
        clear()
        return true
    }

    fun clear() {
        button?.isChecked = false
    }

    fun getAction(): A = action

    private fun contains(position: PositionAPI, x: Float, y: Float): Boolean {
        return x >= position.x &&
            x <= position.x + position.width &&
            y >= position.y &&
            y <= position.y + position.height
    }
}
